pip/pip-385.md

```diff
@@ -0,0 +1,398 @@
+a `ThrottleProducerReceipt` response, we won't be able to read it until further pending messages before that are read.
+
+Here is a sequence diagram highlighting the case when a topic level breach happens:
+
+```mermaid
+%%{init: {"mirrorActors": false, "rightAngles": false} }%%
+sequenceDiagram
+    Client->>Server: CreateProducer(reqId, myTopic)
+    Note right of Server: Check Authorization
+    Server-->>Client: ProducerSuccess(reqId, producerName)
+    Activate Client
+    Activate Server
+    Client->>Server: Send(1, message1)
+    Client->>Server: Send(2, message2)
+    Server-->>Client: SendReceipt(1, msgId1)
+    Note right of Server: Check Throttling
+
+    opt topic/connection throttled
+        Server->>Client: ThrottleProducer(reqId, throttleData)
+        alt client sends receipt in time
+            Client-->>Server: ThrottleProducerReceipt(reqId)
+            note right of Client: client pauses till specified time
+            Note over Client, Server: After some time
+            Client->>Server: Send(3, message3)
+            Client->>Server: Send(4, message4)
+
+        else no response
+            Activate Server
+            note right of Server: TCP read paused
+            Client-xServer: Send(3, message3)
+            Note over Client, Server: After some time
+            opt topic/connection unthrottled
+                note right of Server: TCP read resumed
+                deactivate Server
+                Server-->>Server: read message 3
+                Client->>Server: Send(4, message4)
+            end
+        end
+    end
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Server-->>Client: SendReceipt(4,msgId4)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Java client changes
+
+* Client will now have logic to understand the `ThrottleProducer` command and take relevant action of blocking further
+  messages for the relevant topic. It will then respond back with `ThrottleProducerReceipt` command.
+    * Client will resume message sending after the specified time in the `ThrottleProducer` command's data.
+    * This interval of no messages will be noted as "being throttled"
+    * Within this duration, another `ThrottleProducer` command from server may come.
+* Producer will record new OTel metric indicating which topic was throttled and the reason.
+* In case a message fails due to timeout and there was a throttled command from server for the owning topic, client will
+  instead throw a rate limit exception instead of timeout exception.
+
+# Detailed Design
+
+## High-level Implementation Details
+
+### Broker Changes
+
+* For calls arising from `PublishRateLimiterImpl` class, add logic in `ServerCnxThrottleTracker.java` to send the
+  command to client and wait for response for the max configured duration before calling `changeAutoRead`. It checks for
+  feature availability first.
```

Comment:
> This logic might need some more careful thought.
> 
> It feels that in the case that the client supports the throttling command, that the broker should be able to accept that the rate goes over the limit in a shorter period of time and if this continues for too long, the throttling by changing auto read would have to happen since the broker could run out of memory if this doesn't happen.
> 
> One way to handle this is to send the throttling command to the client as soon as the tokens are below a certain threshold. If the tokens get consumed to a certain value (it can be negative), the broker would have to change auto read to false.
> 
> The flow control will be less accurate when using the throttling command and period based solution. The reason for this is that there's a queue between the client and the broker. When the client stops sending, the broker will continue to receive all produced messages that are in the queue. This is the reason why I was more in favor of a permit (credit) based flow control initially when we started the discussion about this PIP. However, I think that the problem can be solved with the throttling command based solution, but it will most likely require additional details that aren't currently in this PIP. We can figure out the details along the way when we get further with the implementation.

My major concern with permit based (and even if we put in a threshold to send in the command, it kind of falls in same category) is that its a major challenge to implement it correctly and efficiently in a distributed manner. 
The producers here are distributed. We cannot make assumptions on the message rate spread of those producers, specially given how the batching works (round robin) from a client's perspective - leading to sudden bursts to a subset of producers for a topic at any given time. Managing permits in this situation at a sub second level is almost impossible to achieve without massively overshooting or undershooting the rate limiting mark.

I do understand that the producers could snoop in many more messages before they the throttle acknowledgement is read, but doing the other way round also is not feasible i.e. blocking the channel read until all throttle receipts are received ... as the channel is paused...

> If the broker would know how many messages there are in the queues between the client and the broker, it would be possible for the broker to take this into account when calculating the throttling period and when it should send out the command to throttle the producer. However, this would most likely require a separate control channel between the broker and the client to pass this information.
> 
> It is also possible that this problem makes the throttling command useless for producer flow control without the need to constantly use the TCP level throttling. This would make this PIP not meet the original goals.
> 
> There are many ways around the challenges, but solving this aspect is perhaps the hardest problem in implementing this PIP so that it meets the original goals.
> 
> One possible solution could be to estimate the queue size by using the Pulsar's Ping/Pong commands. For this purpose, the broker would have to send Ping commands frequently and measure the round trip time from the Pong responses. However, [the current CommandPing and CommandPong messages don't currently contain any fields](https://github.com/apache/pulsar/blob/9a97c843a46e23a0811e2172991cd00a3af642c0/pulsar-common/src/main/proto/PulsarApi.proto#L701-L704), so passing a request id for correlating the Pong response with the Ping request isn't currently possible in the protocol. It would be necessary to add the `request_id` field as there is for other commands to support using Ping/Pong for calculating RTT.
> 
> The way how RTT is useful in preventing the need to block the reads (by changing auto read to false) is that the broker could assume that the client continues to produce messages with a constant rate and so that it sends the throttle command early enough that the threshold to block reads isn't met. I don't currently see other simple solutions to solve this problem. Perhaps there are also others. I think I had this RTT solution in mind when I first accepted that we can move forward with the throttle command based flow control instead of using permits based flow control. The permits based flow control has other problems so it might eventually be easier to solve flow control with the throttle command based flow control. In the permits based flow control, one of the challenges is that there could be multiple producers and if permits have been sent out to one producer, it could potentially prevent other producers from producing. It's also hard to integrate the permits based flow control to multiple ways of throttling at multiple levels. That's why I think that throttle command based flow control is a better way forward, as long as we are able to solve the challenge that I hopefully described well enough in this comment.

RTT time is one aspect that can help predicting and anticipating the throttling, but there are other aspects as well. TCP is dual channel, but in each way (sending the throttle command, and receiving back the ack) there could be pending messages. Although, the processing of those messages is async, so clearing any pending message read (on broker) or send receipt read (on client side) and reaching the actual throttle/throttle receipt command should be quick enough. This adds to the the network based RTT.

Moreover, calculating RTT using ping/pong would actually also face similar challenge based on the noisiness and variability of the producer rate from the producer in any point in time.. (batches and round robin partition assignment leading to burst patterns at a sub-second level)

This part of the proposal definitely needs more work to improve things, but I think its a good starting point with the pros overtaking the cons.

Reply:
> My major concern with permit based (and even if we put in a threshold to send in the command, it kind of falls in same category) is that its a major challenge to implement it correctly and efficiently in a distributed manner. The producers here are distributed. We cannot make assumptions on the message rate spread of those producers, specially given how the batching works (round robin) from a client's perspective - leading to sudden bursts to a subset of producers for a topic at any given time. Managing permits in this situation at a sub second level is almost impossible to achieve without massively overshooting or undershooting the rate limiting mark.
> 
> I do understand that the producers could snoop in many more messages before they the throttle acknowledgement is read, but doing the other way round also is not feasible i.e. blocking the channel read until all throttle receipts are received ... as the channel is paused...

I agree, permits based solution would be challenging too. However, the challenges could be solved there more accurately since the broker could send the client more specific information about the limits. In Pulsar rate limits are expressed in both messages count and size (bytes). We don't have to go into the solution details here and we can focus on addressing the challenges in the throttling period based solution.


> RTT time is one aspect that can help predicting and anticipating the throttling, but there are other aspects as well. TCP is dual channel, but in each way (sending the throttle command, and receiving back the ack) there could be pending messages. Although, the processing of those messages is async, so clearing any pending message read (on broker) or send receipt read (on client side) and reaching the actual throttle/throttle receipt command should be quick enough. This adds to the the network based RTT.
> 
> Moreover, calculating RTT using ping/pong would actually also face similar challenge based on the noisiness and variability of the producer rate from the producer in any point in time.. (batches and round robin partition assignment leading to burst patterns at a sub-second level)
> 
> This part of the proposal definitely needs more work to improve things, but I think its a good starting point with the pros overtaking the cons.

The RTT could be calculated continuously so that the broker would be able to estimate how long it takes until the client pauses delivery when a throttle command is sent to the client. I don't see any specific problem in calculating the RTT. If there's noisiness, that impacts also the commands. With timestamps, it's possible to find out what the delay is in both directions. The clock differences of the broker and the client can be calculated initially when the connection gets established, before the connection is congested. This all would be a sufficiently accurate estimation of the clock difference and after this, it's possible to calculate how long it takes to travel from broker to client and from client to broker. This would be changing all the time and that's why the Ping/Pong messages would have to be flowing all the time when the connection isn't idling.

The reason why this information is needed is that without the information, the TCP level throttling would be needed. The primary goal of PIP-385 is to avoid that. The way how it could be avoided is sending the throttle producer command from the broker to the client early enough so that the client receives it on time so that it can impact the traffic. 

There would have to be some level of allowance for bursting and that's why the throttling command solution will have a higher peak memory consumption. The peak memory usage can be reduced by improving the algorithm and the prediction logic. The simplest version would assume that the same message rate (count & bytes) is constant and using the RTT information, it could predict how long it would have to ask the client to throttle. 

An improvement could be to have specific fields in the throttle producer response to include client side producer queue information of how many messages and what total size there is waiting to be sent to the broker. Metrics about the current rate could also help on the broker side to adjust the following throttle commands it sends to the client.

I think it's necessary to handle this before PIP-385 could be useful and prevent TCP level throttling in taking place.

Reply:
I looked at many scenarios of extreme cases and average cases and RTT does seem to improve the situation about micro-bursting while we are waiting for throttle command receipt.
Although, I think calculating the RTT would need many more changes in the protocol, so I am proposing that I will raise a followup PIP for it, since RTT isn't specifically for this use case but can be used at other places as well.
What are your thoughts?

Implementation wise, we can ensure that these two PIP's land together, or are cherry-picked to be in the same version if needed.

Reply:
> Although, I think calculating the RTT would need many more changes in the protocol, so I am proposing that I will raise a followup PIP for it, since RTT isn't specifically for this use case but can be used at other places as well.
What are your thoughts?

I think that calculating RTT is relevant to be able to make PIP-385 useful. That's why I don't see a problem in introducing protocol changes for that purpose in the same PIP.

Reply:
> > Although, I think calculating the RTT would need many more changes in the protocol, so I am proposing that I will raise a followup PIP for it, since RTT isn't specifically for this use case but can be used at other places as well.
> > What are your thoughts?
> 
> I think that calculating RTT is relevant to be able to make PIP-385 useful. That's why I don't see a problem in introducing protocol changes for that purpose in the same PIP.

Honestly, I have opposing views here. PIP-385 will improve situation what-so-ever, with or without RTT.
I've already gotten 3 votes here. If I introduce another major protocol change as part of the same PIP, I will have to re-initiate dsicussion and voting.

Honestly, with the massive delays in review, I am opposed to landing massive changes at once.

If you strongly believe that RTT is a _must have_ (which I don't), then the implementation can be planned accordingly, but I do not see a point in increasing the scope of this PIP as it has already gotten enough discussion and votes.

Reply:
> Honestly, I have opposing views here. PIP-385 will improve situation what-so-ever, with or without RTT.
> I've already gotten 3 votes here. If I introduce another major protocol change as part of the same PIP, I will have to re-initiate dsicussion and voting.

It's possible to make progress incrementally too. I'm not blocking you of making progress. The current PIP-385 doesn't currently describe how the rate limiting is achieved in a way how the buffers on the broker side stay within limits so that broker doesn't have to use TCP throttling to stop buffer growth. You can choose to ignore that problem for now, but it's something that we will have to address sooner or later. Incremental approaches are very useful so that we don't have to speculate about possible issues. At the same time, there needs to be a sufficient test cases where the PIP-385 solution can show that buffers stay within limits without TCP throttling. If that's not met, the solution will be useless. It's always good to have a failing test case to guide the implementation. Practice is usually a better teacher than long theoretical discussions. We can come back to these challenges when it becomes relevant in the implementation.

> If you strongly believe that RTT is a must have (which I don't), then the implementation can be planned accordingly, but I do not see a point in increasing the scope of this PIP as it has already gotten enough discussion and votes.

I'm not imposing any restrictions on the implementation as long as the success criteria for PIP-385 is met. I think that RTT will be useful in finding a solution for making PIP-385 successful. I could always be wrong with my gut feeling.